cli/NetworkIO.hpp

```diff
@@ -0,0 +1,286 @@
+      }
+      // While we have a service lock, we add to the Queue. Note that we keep the service lock to protect ourselves from
+      // a race condition in the kill() method.
+      request_state.reset(new RequestState(request));
+
+      // Pushing to the queue will notify consumers.
+      request_queue_.push(request_state.get());
+    }
+    DCHECK(request_state);
+
+    // We have pushed to the request queue, so all there is to do now is wait for Quickstep to process the request.
+    request_state->waitForResponse();
+    if (request_state->getCanceled()) {
+      return Status::CANCELLED;
+    }
+    *response = request_state->getResponse();
+    return Status::OK;
+  }
+
+  /**
+   * @brief The consumer thread waits for a request to materialize.
+   * @return A non-owned RequestState.
+   */
+  RequestState* waitForRequest() {
+    return request_queue_.popOne();
+  }
+
+  /**
+   * @brief Stops accepting further requests and cancels all pending requests.
+   */
+  void kill() {
+    {
+      // This action guarantees that no further requests are added to the queue.
+      std::unique_lock<Mutex> lock(service_mtx_);
+      running_ = false;
+    }
+    // Go through each pending request, and cancel them.
+    while (!request_queue_.empty()) {
+      request_queue_.popOne()->cancel();
+    }
+  }
+
+ private:
+  Mutex service_mtx_;
+  bool running_;
+  ThreadSafeQueue<RequestState*> request_queue_;
+
+  DISALLOW_COPY_AND_ASSIGN(NetworkCliServiceImpl);
+};
+
+class NetworkIOHandle : public IOHandle {
+ public:
+  explicit NetworkIOHandle(RequestState* state)
+      : request_state_(state) {}
+
+  ~NetworkIOHandle() override {
+    // All the commands from the last network interaction have completed, return our response.
+    // This signals to the producer thread that the interaction is complete.
+    request_state_->responseReady(out_stream_.str(), err_stream_.str());
+  }
+
+  FILE* out() override {
+    return out_stream_.file();
+  }
+
+  FILE* err() override {
+    return err_stream_.file();
+  }
+
+  std::string getCommand() override {
+    return request_state_->getRequest();
+  }
+
+ private:
+  MemStream out_stream_, err_stream_;
+  RequestState* request_state_;
+
+  DISALLOW_COPY_AND_ASSIGN(NetworkIOHandle);
+};
+
+/**
+ * A network interface that uses gRPC to accept commands.
+ */
+class NetworkIO : public IOInterface {
+ public:
+  NetworkIO() {
+    grpc::ServerBuilder builder;
+    builder.AddListeningPort(GetAddress(), grpc::InsecureServerCredentials());
+    builder.RegisterService(&service_);
+    server_ = builder.BuildAndStart();
+    LOG(INFO) << "Listening on " << GetAddress();
+  }
+
+  ~NetworkIO() {
+    service_.kill();
+    server_->Shutdown();
+    server_->Wait();
+  }
+
+  IOHandle* getNextIOHandle() override {
+    return new NetworkIOHandle(service_.waitForRequest());
+  }
+
+  /**
+   * @return The underlying service which interacts with gRPC.
+   */
+  NetworkCliServiceImpl& getService() {
```

Comment:
If it is only used by a test, move this method to private, and add that test as a friend.